In [15]:
import pandas as pd
# Load the dataset
df = pd.read_csv('E:\P288_Clustering_Project\P288_Group1_Clustering_Project_Excelr\Dataset\IQR.csv')

In [16]:
df

CO2 Emissions  Days to Start Business  Energy Usage           GDP  \
0           87931.0                  26.000     26998.000  5.479006e+10   
1            9542.0                  26.000      7499.000  9.129595e+09   
2            1617.0                  26.000      1983.000  2.359122e+09   
3            4276.0                  26.000      1836.000  5.788312e+09   
4            1041.0                  26.000     14338.000  2.610959e+09   
...             ...                     ...           ...           ...   
2699         8529.0                  26.000     14338.000  1.507975e+10   
2700         8529.0                   5.000     46913.625  2.213750e+11   
2701         8529.0                   7.000     14338.000  5.000435e+10   
2702         8529.0                  55.625     14338.000  2.213750e+11   
2703         8529.0                  26.000     14338.000  1.507975e+10   

      Health Exp % GDP  Health Exp/Capita  Infant Mortality Rate  \
0                0.035             60.000                 0.0340   
1                0.034             22.000                 0.1075   
2                0.043             15.000                 0.0900   
3                0.047            152.000                 0.0540   
4                0.051             12.000                 0.0960   
...                ...                ...                    ...   
2699             0.061            193.000                 0.0200   
2700             0.118           1291.125                 0.0060   
2701             0.089           1291.125                 0.0100   
2702             0.046            593.000                 0.0130   
2703             0.061            193.000                 0.0200   

      Lending Interest  Life Expectancy Female  Life Expectancy Male  ...  \
0               0.1000                    71.0                  67.0  ...   
1               0.2265                    47.0                  44.0  ...   
2               0.1200                    57.0                  53.0  ...   
3               0.1550                    52.0                  49.0  ...   
4               0.1200                    52.0                  49.0  ...   
...                ...                     ...                   ...  ...   
2699            0.1200                    75.0                  69.0  ...   
2700            0.0330                    81.0                  76.0  ...   
2701            0.1120                    80.0                  74.0  ...   
2702            0.1640                    78.0                  72.0  ...   
2703            0.1200                    83.0                  76.0  ...   

      Population 15-64  Population 65+  Population Total  Tourism Inbound  \
0               0.6190           0.039      3.171945e+07     1.020000e+08   
1               0.4990           0.025      1.392493e+07     3.400000e+07   
2               0.5170           0.029      6.949366e+06     7.700000e+07   
3               0.5870           0.029      1.755375e+06     2.270000e+08   
4               0.5050           0.028      1.160794e+07     2.300000e+07   
...                ...             ...               ...              ...   
2699            0.6425           0.052      3.242700e+04     6.885000e+08   
2700            0.6670           0.136      4.903473e+07     6.341000e+09   
2701            0.6390           0.140      3.395253e+06     2.222000e+09   
2702            0.6520           0.060      2.995478e+07     9.040000e+08   
2703            0.6410           0.152      1.052750e+05     6.885000e+08   

      Tourism Outbound  Birth Rate                   Country Internet Usage  \
0         1.930000e+08       0.020                   Algeria            0.0   
1         1.460000e+08       0.050                    Angola            0.0   
2         5.000000e+07       0.043                     Benin            0.0   
3         2.090000e+08       0.027                  Botswana            0.0   
4         3.000000e+07       0.046         

In [30]:
from sklearn.cluster import KMeans
from sklearn.preprocessing import StandardScaler

# Assuming 'Country' is the categorical column
numerical_cols = df.columns[df.dtypes != 'object'].values

scaler = StandardScaler()
df_scaled = scaler.fit_transform(df[numerical_cols])

"""# Perform KMeans clustering
kmeans = KMeans(n_clusters=5)
clusters = kmeans.fit_predict(df_scaled)"""
agg = AgglomerativeClustering(n_clusters=3)
labels_agg = agg.fit_predict(df_scaled)

# Add the cluster labels to the dataframe as a new feature
df['Cluster'] = clusters

print(df.head())


   CO2 Emissions  Days to Start Business  Energy Usage           GDP  \
0        87931.0                    26.0       26998.0  5.479006e+10   
1         9542.0                    26.0        7499.0  9.129595e+09   
2         1617.0                    26.0        1983.0  2.359122e+09   
3         4276.0                    26.0        1836.0  5.788312e+09   
4         1041.0                    26.0       14338.0  2.610959e+09   

   Health Exp % GDP  Health Exp/Capita  Infant Mortality Rate  \
0             0.035               60.0                 0.0340   
1             0.034               22.0                 0.1075   
2             0.043               15.0                 0.0900   
3             0.047              152.0                 0.0540   
4             0.051               12.0                 0.0960   

   Lending Interest  Life Expectancy Female  Life Expectancy Male  ...  \
0            0.1000                    71.0                  67.0  ...   
1            0.2265         

In [29]:
from sklearn.cluster import KMeans, AgglomerativeClustering, DBSCAN
from sklearn import metrics

# Assuming 'df_normalized' is your DataFrame

# K-Means
kmeans = KMeans(n_clusters=5)
labels_kmeans = kmeans.fit_predict(df_scaled)

# Agglomerative Clustering
agg = AgglomerativeClustering(n_clusters=3)
labels_agg = agg.fit_predict(df_scaled)

# DBSCAN
dbscan = DBSCAN(eps=0.3)
labels_dbscan = dbscan.fit_predict(df_scaled)

# Calculate metrics
for name, labels in zip(['K-Means', 'Agglomerative', 'DBSCAN'], [labels_kmeans, labels_agg, labels_dbscan]):
    print(f"\n{name}:")
    print(f"Silhouette Score: {metrics.silhouette_score(df_scaled, labels)}")
    print(f"Calinski-Harabasz Index: {metrics.calinski_harabasz_score(df_scaled, labels)}")
    print(f"Davies-Bouldin Index: {metrics.davies_bouldin_score(df_scaled, labels)}")


c:\Users\nazee\AppData\Local\Programs\Python\Python311\Lib\site-packages\sklearn\cluster\_kmeans.py:1412: FutureWarning:

The default value of `n_init` will change from 10 to 'auto' in 1.4. Set the value of `n_init` explicitly to suppress the warning




K-Means:
Silhouette Score: 0.22623945940852522
Calinski-Harabasz Index: 1074.8097974850332
Davies-Bouldin Index: 1.446776527742379

Agglomerative:
Silhouette Score: 0.26797203261079106
Calinski-Harabasz Index: 1258.1943544223623
Davies-Bouldin Index: 1.2993235430484733

DBSCAN:
Silhouette Score: -0.36361054172644736
Calinski-Harabasz Index: 4.075530344037714
Davies-Bouldin Index: 1.7659704926628441


In [31]:
import plotly.express as px

# Assuming that df is your DataFrame and it includes a 'Cluster' column
fig = px.scatter(df, x=df.index, y='Cluster', color='Cluster')
fig.show()


In [32]:
import plotly.graph_objects as go

# Filter the DataFrame based on the cluster
cluster0 = df[df['Cluster'] == 0]
cluster1 = df[df['Cluster'] == 1]
cluster2 = df[df['Cluster'] == 2]
cluster3 = df[df['Cluster'] == 3]
cluster4 = df[df['Cluster'] == 4]

# Create a table for each cluster
fig0 = go.Figure(data=[go.Table(header=dict(values=list(cluster0.columns)),
                 cells=dict(values=[cluster0[col] for col in cluster0.columns]))])
fig0.show()

fig1 = go.Figure(data=[go.Table(header=dict(values=list(cluster1.columns)),
                 cells=dict(values=[cluster1[col] for col in cluster1.columns]))])
fig1.show()

fig2 = go.Figure(data=[go.Table(header=dict(values=list(cluster2.columns)),
                 cells=dict(values=[cluster2[col] for col in cluster2.columns]))])
fig2.show()

fig3 = go.Figure(data=[go.Table(header=dict(values=list(cluster3.columns)),
                 cells=dict(values=[cluster3[col] for col in cluster3.columns]))])
fig3.show()

fig4 = go.Figure(data=[go.Table(header=dict(values=list(cluster4.columns)),
                 cells=dict(values=[cluster4[col] for col in cluster4.columns]))])
fig4.show()


In [33]:
import plotly.graph_objects as go

# Filter the DataFrame based on the cluster
cluster0 = df[df['Cluster'] == 0]['Country']
cluster1 = df[df['Cluster'] == 1]['Country']
cluster2 = df[df['Cluster'] == 2]['Country']
cluster3 = df[df['Cluster'] == 3]['Country']
cluster4 = df[df['Cluster'] == 4]['Country']

# Create a table for each cluster
fig0 = go.Figure(data=[go.Table(header=dict(values=['Country']),
                 cells=dict(values=[cluster0]))])
fig0.show()

fig1 = go.Figure(data=[go.Table(header=dict(values=['Country']),
                 cells=dict(values=[cluster1]))])
fig1.show()

fig2 = go.Figure(data=[go.Table(header=dict(values=['Country']),
                 cells=dict(values=[cluster2]))])
fig2.show()

fig3 = go.Figure(data=[go.Table(header=dict(values=['Country']),
                 cells=dict(values=[cluster3]))])
fig3.show()

fig4 = go.Figure(data=[go.Table(header=dict(values=['Country']),
                 cells=dict(values=[cluster4]))])
fig4.show()


In [34]:
import plotly.express as px

# Count the number of countries in each cluster
cluster_counts = df['Cluster'].value_counts().sort_values(ascending=False)

# Create a bar plot
fig = px.bar(cluster_counts, x=cluster_counts.index, y=cluster_counts.values, labels={'x':'Cluster', 'y':'Count'})
fig.show()


In [35]:
import plotly.express as px

# Count the number of countries in each cluster
country_counts = df.groupby('Cluster')['Country'].nunique()

# Create a bar plot
fig = px.bar(country_counts, x=country_counts.index, y=country_counts.values, labels={'x':'Cluster', 'y':'Count'})
fig.show()


In [36]:
# Assuming 'Country' is the categorical column
numerical_cols = df.columns[df.dtypes != 'object'].values

# Calculate the mean value of each numerical feature for each cluster
cluster_characteristics = df.groupby('Cluster')[numerical_cols].mean()

print(cluster_characteristics)


         CO2 Emissions  Days to Start Business  Energy Usage           GDP  \
Cluster                                                                      
0         14729.094156               23.774756  11993.399756  3.014561e+10   
1         64963.771858               17.919057  36631.521175  2.065853e+11   
2          5452.410426               33.821953  11514.960720  1.089380e+10   
3         90304.942308               30.285657  42644.811298  1.602033e+11   
4         12459.984225               30.733882  10226.279321  1.344249e+10   

         Health Exp % GDP  Health Exp/Capita  Infant Mortality Rate  \
Cluster                                                               
0                0.063875         619.224838               0.011576   
1                0.087298        1181.549180               0.005230   
2                0.058461          62.578561               0.072021   
3                0.050885         237.368590               0.026413   
4                0.059598  

The output you've shared shows the mean values of each feature for each cluster. This can be interpreted as the "centroid" or the "characteristic" of each cluster. Here's a brief explanation:

- Cluster 0: This cluster has countries with relatively low CO2 emissions, GDP, and energy usage. The average days to start a business is around 24. The health expenditure as a percentage of GDP is around 0.064 and the average health expenditure per capita is 619. The infant mortality rate is relatively low at 0.012 and the lending interest is 0.103. The life expectancy for females and males is 78 and 73 years respectively.

- Cluster 1: This cluster represents countries with higher CO2 emissions, GDP, and energy usage compared to Cluster 0. The average days to start a business is around 18. The health expenditure as a percentage of GDP is higher at 0.087 and the average health expenditure per capita is significantly higher at 1182. The infant mortality rate is lower at 0.005 and the lending interest is lower at 0.079. The life expectancy for females and males is significantly higher at 82 and 77 years respectively.

- Cluster 2: This cluster represents countries with even lower CO2 emissions, GDP, and energy usage compared to Cluster 0. The average days to start a business is around 34. The health expenditure as a percentage of GDP is slightly lower at 0.058 and the average health expenditure per capita is the lowest at 63. The infant mortality rate is higher at 0.072 and the lending interest is higher at 0.157. The life expectancy for females and males is significantly lower than other clusters at 56 and 54 years respectively.

- Cluster 3: This cluster represents countries with high CO2 emissions, GDP, and energy usage but lower than Cluster 1. The average days to start a business is around 30. The health expenditure as a percentage of GDP is lower at 0.051 and the average health expenditure per capita is relatively low at 237. The infant mortality rate is relatively low at 0.026 and the lending interest is slightly higher at 0.125 compared to Cluster 1 but lower than Cluster 2. The life expectancy for females and males is in between other clusters at around 73 and 67 years respectively.

- Cluster 4: This cluster represents countries with low CO2 emissions, GDP, and energy usage similar to Cluster 0 but slightly higher on average. The average days to start a business is around 31 which is higher than Cluster 0 but lower than others. The health expenditure as a percentage of GDP is slightly lower than Cluster 0 at around 0.060 and the average health expenditure per capita is relatively low at around 182 which is higher than Cluster 2 but lower than others. The infant mortality rate is relatively low at around 0.025 which is similar to Cluster 3 but lower than others, and the lending interest rate is slightly higher than others except for Cluster2 at around .140 . The life expectancy for females and males are similar to Cluster3.

Please note that this interpretation assumes that all features have equal importance in defining the clusters, which might not be the case depending on your specific use case or the nature of your data.

The output you've shared shows the mean values of each feature for each cluster. This can be interpreted as the "centroid" or the "characteristic" of each cluster. Here's a brief explanation:

- Cluster 0: This cluster has countries with relatively low CO2 emissions, GDP, and energy usage. The average days to start a business is around 24. The health expenditure as a percentage of GDP is around 0.064 and the average health expenditure per capita is 619. The infant mortality rate is relatively low at 0.012 and the lending interest is 0.103. The life expectancy for females and males is 78 and 73 years respectively.

- Cluster 1: This cluster represents countries with higher CO2 emissions, GDP, and energy usage compared to Cluster 0. The average days to start a business is around 18. The health expenditure as a percentage of GDP is higher at 0.087 and the average health expenditure per capita is significantly higher at 1182. The infant mortality rate is lower at 0.005 and the lending interest is lower at 0.079. The life expectancy for females and males is significantly higher at 82 and 77 years respectively.

- Cluster 2: This cluster represents countries with even lower CO2 emissions, GDP, and energy usage compared to Cluster 0. The average days to start a business is around 34. The health expenditure as a percentage of GDP is slightly lower at 0.058 and the average health expenditure per capita is the lowest at 63. The infant mortality rate is higher at 0.072 and the lending interest is higher at 0.157. The life expectancy for females and males is significantly lower than other clusters at 56 and 54 years respectively.

- Cluster 3: This cluster represents countries with high CO2 emissions, GDP, and energy usage but lower than Cluster 1. The average days to start a business is around 30. The health expenditure as a percentage of GDP is lower at 0.051 and the average health expenditure per capita is relatively low at 237. The infant mortality rate is relatively low at 0.026 and the lending interest is slightly higher at 0.125 compared to Cluster 1 but lower than Cluster 2. The life expectancy for females and males is in between other clusters at around 73 and 67 years respectively.

- Cluster 4: This cluster represents countries with low CO2 emissions, GDP, and energy usage similar to Cluster 0 but slightly higher on average. The average days to start a business is around 31 which is higher than Cluster 0 but lower than others. The health expenditure as a percentage of GDP is slightly lower than Cluster 0 at around 0.060 and the average health expenditure per capita is relatively low at around 182 which is higher than Cluster 2 but lower than others. The infant mortality rate is relatively low at around 0.025 which is similar to Cluster 3 but lower than others, and the lending interest rate is slightly higher than others except for Cluster2 at around .140 . The life expectancy for females and males are similar to Cluster3.

Please note that this interpretation assumes that all features have equal importance in defining the clusters, which might not be the case depending on your specific use case or the nature of your data.